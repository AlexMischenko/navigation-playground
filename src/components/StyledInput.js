import React, { useState } from 'react'
import { StyleSheet, View, Text, TextInput } from 'react-native'

import { adjustToWidth, adjustFont } from '../utils/styleHelpers'

const StyledInput = ({
  title,
  style,
  titleStyle,
  textInputStyle,
  validationRegEx,
  onChangeText,
  value,
  numbersOnly,
  error,
  disabled,
  ...props
}) => {
  const [localValue, setLocalValue] = useState(value ? `${value}` : '')
  const showError = error || disabled

  const handleChangeText = (newValue) => {
    if (!validationRegEx || newValue.match(validationRegEx) || newValue === '') {
      setLocalValue(newValue)
      onChangeText(numbersOnly ? parseFloat(newValue) : newValue)
    }
  }

  return (
    <View style={[cs.inputContainer, style]} elevation={5}>
      <Text style={[cs.inputTitleText, cs.titleMargin, titleStyle]}>{title}</Text>
      <View style={[cs.inputBlock, showError && cs.errorBackground]} elevation={5}>
        <TextInput
          style={[cs.input, textInputStyle]}
          {...props}
          value={localValue}
          onChangeText={handleChangeText}
        />
      </View>
    </View>
  )
}

const cs = StyleSheet.create({
  inputContainer: {},
  inputTitleText: {
    fontSize: adjustFont(14),
    color: 'black',
    textTransform: 'capitalize',
  },
  inputBlock: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    width: '100%',
    height: adjustToWidth(64),
    paddingLeft: adjustToWidth(17),
    paddingRight: adjustToWidth(16.5),
    borderRadius: adjustToWidth(16),
    backgroundColor: 'white',
    shadowColor: 'gray',
    shadowOpacity: 0.07,
    shadowRadius: adjustToWidth(10),
    shadowOffset: {
      width: adjustToWidth(0),
      height: adjustToWidth(0),
    },
  },
  input: {
    flex: 1,
  },
  titleMargin: {
    marginBottom: adjustToWidth(9.5),
  },
  errorBackground: {
    borderWidth: adjustToWidth(1),
    borderColor: 'red',
  },
})

export default StyledInput
