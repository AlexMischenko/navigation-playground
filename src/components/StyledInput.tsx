import React, { useState } from 'react'
import { StyleSheet, View, Text, TextInput } from 'react-native'

import type { IStyledInputProps, ITheme } from '@types'
import { useTheme, useStyleSheet } from '@hooks'
import { adjustToWidth } from '@utils/styleHelpers'

const StyledInput: React.FC<IStyledInputProps> = ({
  title,
  style,
  titleStyle,
  textInputStyle,
  validationRegEx,
  onChangeText,
  value,
  numbersOnly,
  error,
  disabled,
  ...props
}) => {
  const [localValue, setLocalValue] = useState(value ? `${value}` : '')
  const theme = useTheme()
  const cs = useStyleSheet(styles)
  const showError = error || disabled

  const handleChangeText = (newValue: string) => {
    if (!validationRegEx || newValue.match(validationRegEx) || newValue === '') {
      setLocalValue(newValue)
      onChangeText(numbersOnly ? parseFloat(newValue) : newValue)
    }
  }

  return (
    <View style={style}>
      <Text style={[cs.inputTitleText, theme.helpers.mb8, titleStyle]}>{title}</Text>
      <View style={[cs.inputBlock, showError && cs.errorBackground]}>
        <TextInput
          {...props}
          style={[theme.helpers.flex1, textInputStyle]}
          value={localValue}
          onChangeText={handleChangeText}
        />
      </View>
    </View>
  )
}

const styles = (theme: ITheme): any =>
  StyleSheet.create({
    inputTitleText: {
      ...theme.textStyles.callout,
      textTransform: 'capitalize',
    },
    inputBlock: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      width: '100%',
      height: adjustToWidth(64),
      paddingLeft: adjustToWidth(17),
      paddingRight: adjustToWidth(16.5),
      borderRadius: theme.borderRadius.lg,
      backgroundColor: theme.colors.white,
      shadowColor: theme.colors.grey,
      shadowOpacity: 0.15,
      shadowRadius: adjustToWidth(10),
      shadowOffset: {
        width: adjustToWidth(0),
        height: adjustToWidth(0),
      },
      elevation: 3,
    },
    errorBackground: {
      borderWidth: StyleSheet.hairlineWidth,
      borderColor: 'red',
    },
  })

export default StyledInput
