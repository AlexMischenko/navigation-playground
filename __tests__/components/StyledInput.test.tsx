import * as React from 'react'
import { render, fireEvent } from '@testing-library/react-native'

import StyledInput from '../../src/components/StyledInput'

describe('Testing @components/StyledInput', () => {
  test('`StyledInput` has title displayed when "title" prop set', () => {
    const component = <StyledInput title="TestTitle" onChangeText={jest.fn()} />
    const { getByText } = render(component)
    const titleText = getByText('TestTitle')

    expect(titleText).toBeTruthy()
  })

  test('`StyledInput` passes numbers in "onChangeText" when "numbersOnly" prop enabled', () => {
    const onChangeTextMock = jest.fn()
    const component = <StyledInput placeholder="ph" onChangeText={onChangeTextMock} numbersOnly />
    const { getByPlaceholderText } = render(component)
    const textInput = getByPlaceholderText('ph')

    fireEvent.changeText(textInput, '1234')

    expect(onChangeTextMock).toHaveBeenCalledWith(1234)
  })

  test('`StyledInput` doesn`t call "onChangeText" when value doesn`t match "validationRegEx"', () => {
    const onChangeTextMock = jest.fn()
    const component = (
      <StyledInput placeholder="ph" onChangeText={onChangeTextMock} validationRegEx={/^[a-z]*$/} />
    )
    const { getByPlaceholderText } = render(component)
    const textInput = getByPlaceholderText('ph')

    fireEvent.changeText(textInput, 'QWERTY')

    expect(onChangeTextMock).not.toHaveBeenCalledWith('QWERTY')
  })

  test('`StyledInput` displays error state when "error" prop enabled', () => {
    const onChangeTextMock = jest.fn()
    const component = <StyledInput onChangeText={onChangeTextMock} error />
    const wrapper = render(component)

    expect(wrapper).toMatchSnapshot()
  })
})
