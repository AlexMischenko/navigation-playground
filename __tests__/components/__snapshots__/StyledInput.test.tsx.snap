// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing @components/StyledInput \`StyledInput\` displays error state when "error" prop enabled 1`] = `
<View>
  <Text
    style={
      Array [
        Object {
          "color": "#2B2D42",
          "fontSize": 28.985507246376812,
          "letterSpacing": -0.5797101449275363,
          "lineHeight": 38.04347826086956,
          "textTransform": "capitalize",
        },
        Object {
          "marginBottom": 14.5,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderRadius": 25.5,
          "elevation": 3,
          "flexDirection": "row",
          "height": 116,
          "justifyContent": "space-between",
          "paddingLeft": 31,
          "paddingRight": 30,
          "shadowColor": "#8D99AE",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.15,
          "shadowRadius": 18,
          "width": "100%",
        },
        Object {
          "borderColor": "red",
          "borderWidth": 0.5,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;
